services/device-registry-mongodb/src/main/java/org/eclipse/hono/deviceregistry/mongodb/service/MongoDbBasedTenantService.java

```diff
@@ -0,0 +1,572 @@
+     *             and use this span as the parent for any spans created in this method.
+     * @return A future indicating the outcome of the operation.
+     *         The <em>status</em> will be
+     *         <ul>
+     *         <li><em>200 OK</em> if a tenant with the given ID is registered. The <em>payload</em> will contain the
+     *         tenant's configuration information.</li>
+     *         <li><em>404 Not Found</em> if no tenant with the given identifier and version exists.</li>
+     *         </ul>
+     * @throws NullPointerException if {@code tenantId} is {@code null}.
+     */
+    private Future<OperationResult<Tenant>> processReadTenant(final String tenantId, final Span span) {
+        Objects.requireNonNull(tenantId);
+
+        return findTenant(tenantId)
+                .compose(tenantDto -> Future.succeededFuture(OperationResult.ok(
+                        HttpURLConnection.HTTP_OK,
+                        tenantDto.getTenant(),
+                        Optional.ofNullable(DeviceRegistryUtils.getCacheDirective(config.getCacheMaxAge())),
+                        Optional.ofNullable(tenantDto.getVersion()))));
+
+    }
+
+    /**
+     * Gets a tenant dto by tenant id.
+     *
+     * @param tenantId the tenant id.
+     * @return A future indicating the outcome of the operation.
+     *         <p>
+     *         The future with a {@link TenantDto} be provided if found.
+     *         Otherwise the future will fail with
+     *         a {@link ClientErrorException} with {@link HttpURLConnection#HTTP_NOT_FOUND}.
+     * @throws NullPointerException if the parameters is {@code null}.
+     */
+    private Future<TenantDto> findTenant(final String tenantId) {
+        Objects.requireNonNull(tenantId);
+
+        final JsonObject findTenantQuery = new MongoDbDocumentBuilder()
+                .withTenantId(tenantId)
+                .document();
+        final Promise<JsonObject> didReadTenant = Promise.promise();
+        mongoClient.findOne(getConfig().getCollectionName(), findTenantQuery, new JsonObject(), didReadTenant);
+        return didReadTenant.future()
+                .compose(successDidReadTenant -> Optional.ofNullable(successDidReadTenant)
+                        .map(tenant -> tenant.mapTo(TenantDto.class))
+                        .map(Future::succeededFuture)
+                        .orElseGet(() -> Future.failedFuture(new ClientErrorException(HttpURLConnection.HTTP_NOT_FOUND,
+                                String.format("Tenant [%s] not found.", tenantId)))));
+    }
+
+    @Override
+    public Future<OperationResult<Id>> createTenant(final Optional<String> tenantId, final Tenant tenantObj,
+            final Span span) {
+        Objects.requireNonNull(tenantId);
+        Objects.requireNonNull(tenantObj);
+
+        return MongoDbDeviceRegistryUtils.isAllowedToModify(config, tenantId.orElse(""))
```

Comment:
If tenant identifier is not provided then UUID (`getUniqueIdentifier()`) is to be used.
Also I see, that you have generated tenant Id later in the `processCreateTenant(...)`. It is important that is generated here and then the same  generated id is used further. As the conflict should be checked with the id that has been already generated. After that, you could create a dto and pass it to the `processCreateTenant(...)` similar to the `processCreateDevice(...)`

Reply:
`isAllowedToModify` only uses the `tenantId` for a error message, which is irrelevant if it is a generated one and then dropped. 
`checkConflictingCATenants` needs a `tenantDto` only in `updateTenant`, since it should ignore this tenant do avoid conflict matches with itself. `tenantId` conflict checks are implicit with `mongoClient.insert`.
Nevertheless, `getUniqueIdentifier` be called there instead of in `processCreateTenant` to make it clearer.

Reply:
> `isAllowedToModify` only uses the `tenantId` for a error message, which is irrelevant if it is a generated one and then dropped.

In that case, instead of using an empty string as the tenant id, how about invoking `isAllowedToModify` only when tenant id exists else not.

Reply:
That would be wrong, since the allowance does not depend on the `tenantId`.
`tenantId ` is now generated in `createTenant` before the `isAllowedToModify` call.